source/load-balancers/load-balancers.rst

```diff
@@ -0,0 +1,535 @@
+Handshake MUST also be ignored for SDAM error handling purposes. Once the initial
+handshake is complete, the connection MUST determine its generation number based
+on the :code:`serviceId` field in the handshake response. Any errors that occur
+during the rest of connection establishment (e.g. errors during authentication commands)
+MUST go through the SDAM error handling flow but MUST NOT mark the server as
+:code:`Unknown`.
+
+Post-Handshake Errors
+^^^^^^^^^^^^^^^^^^^^^^
+
+When the driver is operating in load balanced mode and an application operation receives a
+state change error, the driver MUST NOT make any changes to the :code:`TopologyDescription`
+or the :code:`ServerDescription` of the load balancer (i.e. it MUST NOT mark the load
+balancer as :code:`Unknown`). If the error requires the connection pool to be cleared,
+the driver MUST only clear connections with the same :code:`serviceId` as the connection
+which errored.
+
+
+Events
+------
+
+When in load balancer mode the driver MUST now include the :code:`serviceId` in the
+:code:`CommandStartedEvent`, :code:`CommandSucceededEvent`, and
+:code:`CommandFailedEvent`. The driver MAY decide how to expose this information.
+Drivers that have a :code:`ConnectionId` object for example, MAY choose to provide a
+:code:`serviceId` in that object. The :code:`serviceId` field is only present when
+in load balancer mode and connected to a service that is behind a load balancer.
+
+Additionally the :code:`PoolClearedEvent` MUST also contain a :code:`serviceId`
+field.
+
+
+Downstream Visible Behavioral Changes
+-------------------------------------
+
+Services MAY add a command line option or other configuration parameter, that tells the service
+it is running behind a load balancer. Services MAY also dynamically determine whether they are
+behind a load balancer.
+
+When the server supports the :code:`loadBalanced` option, the server MUST respond to the command
+with a :code:`serviceId` field in the result.
+
+All services which terminate TLS MUST be configured to return a TLS certificate for a hostname
+which matches the hostname the client is connecting to.
+
+All services behind a load balancer that have been started with the aforementioned option MUST
+add a top level :code:`serviceId` field to their response to the :code:`hello` command. This field
+MUST be a BSON :code:`ObjectId` and SHOULD NOT change while the service is running.
+
+All services that have the behaviour of reaping idle cursors after a specified period of time MAY
+also close the connection associated with the cursor when the cursor is reaped. Conversely, those
+services MAY reap a cursor when the connection associated with the cursor is closed.
+
+All services that have the behaviour of reaping idle transactions after a specified period of time
+MUST also close the connection associated with the transaction when the transaction is reaped.
```

Comment:
Here, we say the service MUST close the connection when reaping the transaction, but the previous paragraph says the service MAY close the connection when reaping cursors. Should we standardize on MAY? IIRC, these were both previously MUST but we realized that servers may not need to invalidate the connection as part of their implementation.

Reply:
Changed to MAY in this case, nice catch.